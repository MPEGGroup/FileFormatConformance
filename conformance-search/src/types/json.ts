export type Coverage = {
    lists: {
        boxes: {
            covered: Array<string>;
            not_covered: Array<string>;
            under_consideration: Array<string>;
        };
        features: {
            covered: Array<string>;
            not_covered: Array<string>;
            under_consideration: Array<string>;
        };
    };
    path_file_map: {
        count: number;
        non_empty: number;
        overall_coverage: number;
        published: {
            count: number;
            coverage: number;
        };
        under_consideration: {
            count: number;
            coverage: number;
        };
    };
    feature_file_map: {
        count: number;
        published_features: number;
        under_consideration_features: number;
        features_from_spec: number;
        features_from_file: number;
    };
    files: {
        count: number;
        published_count: number;
        under_consideration_count: number;
    };
    boxes: {
        count: number;
    };
    spec_info: Array<{
        ISO: string;
        name: string;
        version: number;
        amendmends?: Array<string>;
        corrigenda?: Array<string>;
    }>;
};

export type FileMetadata = {
    contributor: string;
    description: string;
    license: string;
    md5: string;
    rel_filepath: string;
    abs_filepath: string;
    version: number;
    published: boolean;
    associated_files: string[];
    mdms_number: string;
    features: string[];
    conforms_to: {
        ISO: string;
        version: number;
        amendmends?: Array<string>;
        corrigenda?: Array<string>;
    }[];
    notes:
        | string
        | {
              [key: string]: unknown;
          };
};

export type FeatureFileMap = {
    key: string;
    description: string;
    spec: string;
    added_from_file: "spec" | "file";
    files: string[];
};

export type Files = {
    not_found: string[];
    path_file_map: {
        [path: string]: {
            [variant: string]: string[];
        };
    };
    feature_file_map: {
        [feature: string]: FeatureFileMap;
    };
    file_metadata: {
        [file: string]: FileMetadata;
    };
    boxes: {
        [box: string]: string[][];
    };
};

export type Dictionary = {
    fourccs: string[];
    types: string[];
    specs: string[];
    registered_types: string[];
};

export type Hierarchy = {
    [key: string]: Hierarchy;
};
