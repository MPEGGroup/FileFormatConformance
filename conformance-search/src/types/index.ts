import Fuse from "fuse.js";
import { FileMetadata } from "./json";

export type Box = {
    fourcc: string;
    spec: string;
    codec: string;
    description: string;
    type: string;
    versions: number[];
    flags: (
        | {
              name: string;
              value: string;
              description: string;
          }
        | {
              wildcard: string;
          }
    )[];
    containers: {
        fourcc: string;
        type: string;
    }[];
    syntax: string;
};

export type Feature = {
    name: string;
    description: string;
    spec: string;
};

export type Filter = {
    type: string;
    value: string;
};

export type FilterConverted = {
    [key: string]: string;
};

export type SearchResult<T> = {
    item: T;
    score: number;
    actualScore: number;
    refIndex: number;
};

export type SearchResultRefined<T> = SearchResult<T> & {
    count: number;
    type: string;
    exclude: boolean;
    refinements?: {
        variant: {
            versions: {
                value: number[];
            };
            flags: {
                value: string[];
                exact: boolean;
            };
        };
    };
};

export type FuseSearchWithScore<T> = Fuse.FuseResult<T> & {
    score: number;
    actualScore: number;
};

export type FileSearchResult = {
    item: FileMetadata;
    matches: {
        type: string;
        value: string;
    }[];
};

export type CoverageResults = {
    boxes: {
        covered: Fuse.FuseResult<string>[];
        not_covered: Fuse.FuseResult<string>[];
        under_consideration: Fuse.FuseResult<string>[];
        [key: string]: Fuse.FuseResult<string>[];
    };
    features: {
        covered: Fuse.FuseResult<string>[];
        not_covered: Fuse.FuseResult<string>[];
        under_consideration: Fuse.FuseResult<string>[];
        [key: string]: Fuse.FuseResult<string>[];
    };
};
