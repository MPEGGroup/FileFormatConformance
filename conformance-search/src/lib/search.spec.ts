import _ from "lodash";
import { Box, SearchResult } from "@/types";
import Database from "./database";
import Search from "./search";

describe("Search", () => {
    describe("sanity checks", async () => {
        let search: Search;
        beforeAll(async () => {
            search = await Search.getInstance();
        });

        it("search every box", async () => {
            const allBoxes = (await Database.getInstance()).boxes;

            const notFound: string[] = [];
            // eslint-disable-next-line no-restricted-syntax
            for await (const box of allBoxes) {
                const { boxes } = await search.search(`="${box.fourcc}"`, [
                    {
                        type: "type",
                        value: `="${box.type}"`
                    }
                ]);
                if (box.fourcc !== "file" && boxes.length !== 1) {
                    notFound.push(box.fourcc);
                }
            }

            expect(notFound).toEqual([]);
        });

        it("should not return anything", async () => {
            expect(await search.search("")).toEqual({
                boxes: [],
                features: []
            });
        });
    });

    describe("search of 'meta'", () => {
        let search: Search;
        beforeAll(async () => {
            search = await Search.getInstance();
        });

        it("without any filters", async () => {
            const { boxes } = await search.search("meta");

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "meta" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("with a filter", async () => {
            const { boxes } = await search.search("meta", [{ type: "type", value: "=FullBox" }]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "meta" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.type)).not.toContain("Box");
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("with a filter that has no value", async () => {
            const { boxes } = await search.search("meta", [{ type: "type", value: "" }]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "meta" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("with multiple filters of same type", async () => {
            const { boxes } = await search.search("meta", [
                { type: "type", value: "FullBox" },
                { type: "type", value: "Box" }
            ]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "meta" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });
    });

    describe("search with other filters", () => {
        let search: Search;
        beforeAll(async () => {
            search = await Search.getInstance();
        });

        it("with a spec filter", async () => {
            const { boxes } = await search.search("", [{ type: "spec", value: "14496-12" }]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });
    });

    describe("search with container filter", () => {
        let search: Search;
        beforeAll(async () => {
            search = await Search.getInstance();
        });

        it("$mdat > '=free'", async () => {
            const { boxes } = await search.search("=free", [{ type: "container", value: "$mdat" }]);

            expect(boxes.length).toEqual(1);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "free" && r.score === 0)
            ).toBeTruthy();
        });

        it("$trak > 'encf'", async () => {
            const { boxes } = await search.search("encf", [{ type: "container", value: "$trak" }]);

            expect(boxes.length).toEqual(1);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "encf" && r.score === 0)
            ).toBeTruthy();
        });

        it("$avcC > ''", async () => {
            const { boxes } = await search.search("", [{ type: "container", value: "$avcC" }]);

            expect(boxes.length).toEqual(3);
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("$file > $meta > 'idat'", async () => {
            const { boxes } = await search.search("idat", [
                { type: "container", value: "$file.$meta" }
            ]);

            expect(boxes.length).toEqual(1);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "idat" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("(exact) $file > $moov > 'trak'", async () => {
            const { boxes } = await search.search("trak", [
                { type: "container", value: "=$file.$moov" }
            ]);

            expect(boxes.length).toEqual(1);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "trak" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("(exact) $moov > $trak > $meta > 'hdlr'", async () => {
            const { boxes } = await search.search("hdlr", [
                { type: "container", value: "=$moov.$trak.$meta" }
            ]);

            expect(boxes.length).toEqual(1);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "hdlr" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("(exact) $trak > ''", async () => {
            const { boxes } = await search.search("", [{ type: "container", value: "=$trak" }]);

            expect(boxes.length).toEqual(11);
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("$moov > $edts > 'elst'", async () => {
            it("exact match", async () => {
                const { boxes } = await search.search("elst", [
                    { type: "container", value: "=$moov.$edts" }
                ]);

                expect(boxes.length).toEqual(0);
            });

            it("normal match", async () => {
                const { boxes } = await search.search("elst", [
                    { type: "container", value: "$moov.$edts" }
                ]);

                expect(boxes.length).toEqual(1);
                expect(
                    _.some(
                        boxes,
                        (r: SearchResult<Box>) => r.item.fourcc === "elst" && r.score === 0
                    )
                ).toBeTruthy();
                expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
            });
        });

        it("$mdat > $free > ''", async () => {
            const { boxes } = await search.search("=free", [
                { type: "container", value: "$mdat.$free" }
            ]);

            expect(boxes.length).toEqual(0);
        });

        it("$moov > $meta > 'xml'", async () => {
            const { boxes } = await search.search("xml", [
                { type: "container", value: "$moov.$meta" }
            ]);

            expect(boxes.length).toEqual(2);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "xml " && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("#VisualSampleEntry > ''", async () => {
            const { boxes } = await search.search("", [
                { type: "container", value: "#VisualSampleEntry" }
            ]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "seii" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });

        it("entering container value one by one", async () => {
            await Database.getInstance(true);
            search = await Search.getInstance(true);

            await search.search("", [{ type: "container", value: "" }]);
            await search.search("", [{ type: "container", value: "=$file" }]);
            await search.search("", [{ type: "container", value: "=$file.$moov" }]);
            const { boxes: result1 } = await search.search("trak", [
                { type: "container", value: "=$file.$moov" }
            ]);

            expect(result1.length).toEqual(1);
            expect(
                _.some(result1, (r: SearchResult<Box>) => r.item.fourcc === "trak" && r.score === 0)
            ).toBeTruthy();

            await search.search("", [{ type: "container", value: "" }]);
            await search.search("", [{ type: "container", value: "=$moov" }]);
            await search.search("", [{ type: "container", value: "=$moov.$trak" }]);
            await search.search("", [{ type: "container", value: "=$moov.$trak.$meta" }]);
            const { boxes: result2 } = await search.search("hdlr", [
                { type: "container", value: "=$moov.$trak.$meta" }
            ]);

            expect(result2.length).toEqual(1);
            expect(
                _.some(result2, (r: SearchResult<Box>) => r.item.fourcc === "hdlr" && r.score === 0)
            ).toBeTruthy();
        });

        it("caches previous calculations", async () => {
            // Just for cache
            await search.search("xml", [{ type: "container", value: "$moov.$meta" }]);

            const { boxes } = await search.search("", [{ type: "container", value: "$moov" }]);

            expect(boxes.length).toBeGreaterThan(0);
            expect(
                _.some(boxes, (r: SearchResult<Box>) => r.item.fourcc === "avcp" && r.score === 0)
            ).toBeTruthy();
            expect(boxes.map((r) => r.item.fourcc)).toMatchSnapshot();
        });
    });
});
