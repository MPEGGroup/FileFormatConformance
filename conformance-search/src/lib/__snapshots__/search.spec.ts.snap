// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > search of 'meta' > with a filter 1`] = `
[
  "meta",
]
`;

exports[`Search > search of 'meta' > with a filter that has no value 1`] = `
[
  "meta",
]
`;

exports[`Search > search of 'meta' > with multiple filters of same type 1`] = `
[
  "meta",
  "leva",
  "mvra",
  "mehd",
  "mstv",
  "ipma",
  "schm",
  "xml ",
  "csch",
  "pitm",
  "uriI",
  "iinf",
  "csgp",
  "stz2",
  "txtC",
  "saio",
  "saiz",
  "mvhd",
  "infe",
  "mfhd",
  "tfhd",
  "trun",
  "mdhd",
  "sthd",
  "tfdt",
  "tfra",
  "stco",
  "dref",
  "mfro",
  "tkhd",
  "hmhd",
  "nmhd",
  "stss",
  "smhd",
  "vmhd",
  "stts",
]
`;

exports[`Search > search of 'meta' > without any filters 1`] = `
[
  "meta",
]
`;

exports[`Search > search with container filter > #SampleEntry > '' 1`] = `
[
  "auxi",
  "btrt",
  "ccst",
  "csch",
  "free",
  "frma",
  "rinf",
  "schi",
  "schm",
  "scrb",
  "seii",
  "sinf",
  "skip",
  "stvi",
  "uuid",
]
`;

exports[`Search > search with container filter > $avcC > '' 1`] = `
[
  "free",
  "skip",
  "uuid",
]
`;

exports[`Search > search with container filter > $file > $meta > 'idat' 1`] = `
[
  "idat",
]
`;

exports[`Search > search with container filter > $moov > $meta > 'xml' 1`] = `
[
  "xml ",
  "bxml",
]
`;

exports[`Search > search with container filter > (exact) $file > $moov > 'trak' 1`] = `
[
  "trak",
]
`;

exports[`Search > search with container filter > (exact) $moov > $trak > $meta > 'hdlr' 1`] = `
[
  "hdlr",
]
`;

exports[`Search > search with container filter > (exact) $moov > '' TBD 1`] = `
[
  "free",
  "meta",
  "mvex",
  "mvhd",
  "skip",
  "trak",
  "udta",
  "uuid",
]
`;

exports[`Search > search with container filter > (exact) $trak > '' 1`] = `
[
  "edts",
  "free",
  "mdia",
  "meta",
  "skip",
  "tkhd",
  "tref",
  "trgr",
  "ttyp",
  "udta",
  "uuid",
]
`;

exports[`Search > search with container filter > caches previous calculations 1`] = `
[
  "3dpr",
  "a3d1",
  "a3d2",
  "a3d3",
  "a3d4",
  "a3dC",
  "acgl",
  "adda",
  "adrc",
  "aebr",
  "aebr",
  "aebr",
  "afbr",
  "afbr",
  "afbr",
  "albc",
  "alou",
  "alst",
  "alte",
  "altr",
  "altt",
  "amgl",
  "assp",
  "aud ",
  "auxC",
  "auxi",
  "auxl",
  "auxl",
  "avc1",
  "avc1",
  "avc2",
  "avc2",
  "avc3",
  "avc3",
  "avc4",
  "avc4",
  "avcC",
  "avcC",
  "avcp",
  "avcp",
  "avll",
  "avss",
  "base",
  "brnd",
  "brst",
  "btrt",
  "buff",
  "bxml",
  "cclv",
  "cclv",
  "ccst",
  "cdsc",
  "chnl",
  "cinf",
  "clap",
  "clap",
  "clli",
  "clli",
  "co64",
  "colr",
  "colr",
  "cprt",
  "crtt",
  "csch",
  "csgp",
  "cslg",
  "cstg",
  "ctts",
  "dcfi",
  "deps",
  "dimg",
  "dimm",
  "dinf",
  "dmax",
  "dmed",
  "dmix",
  "dobr",
  "dobr",
  "dobr",
  "dpnd",
  "drap",
  "dref",
  "drep",
  "dtrt",
  "edts",
  "elng",
  "elst",
  "enc3",
  "enca",
  "enca",
  "encf",
  "encm",
  "encp",
  "encs",
  "enct",
  "encu",
  "encv",
  "eob ",
  "eos ",
  "eqiv",
  "eqiv",
  "etyp",
  "evcr",
  "exbl",
  "favc",
  "fdel",
  "fdel",
  "fdp ",
  "fecr",
  "fiin",
  "fire",
  "fobr",
  "fobr",
  "fobr",
  "font",
  "fpar",
  "free",
  "frma",
  "gitn",
  "grid",
  "grpl",
  "hdlr",
  "hev1",
  "hev1",
  "hev2",
  "hev2",
  "hev3",
  "hev3",
  "hind",
  "hinf",
  "hint",
  "hmhd",
  "hnti",
  "hvc1",
  "hvc1",
  "hvc2",
  "hvc2",
  "hvc3",
  "hvc3",
  "hvcC",
  "hvcC",
  "iaug",
  "icpv",
  "idat",
  "iinf",
  "iloc",
  "iloc",
  "imdt",
  "imir",
  "infe",
  "ipco",
  "ipma",
  "ipro",
  "iprp",
  "iref",
  "iroi",
  "irot",
  "iscl",
  "ispe",
  "jpgC",
  "kind",
  "lbli",
  "ldep",
  "leva",
  "lhe1",
  "lhv1",
  "lhvC",
  "lhvC",
  "linf",
  "lsel",
  "ludt",
  "maxr",
  "md5i",
  "mdcv",
  "mdcv",
  "mdft",
  "mdhd",
  "mdia",
  "mebx",
  "mehd",
  "meta",
  "mett",
  "metx",
  "minf",
  "minp",
  "mixn",
  "msrc",
  "mstv",
  "mvc1",
  "mvc2",
  "mvc3",
  "mvc4",
  "mvcC",
  "mvcg",
  "mvci",
  "mvd1",
  "mvd2",
  "mvd3",
  "mvd4",
  "mvdC",
  "mvex",
  "mvhd",
  "mvif",
  "mvra",
  "nalm",
  "nmhd",
  "npck",
  "nump",
  "oinf",
  "oinf",
  "opeg",
  "opth",
  "oref",
  "padb",
  "paen",
  "pano",
  "pano",
  "pano",
  "pase",
  "pasp",
  "pasp",
  "payt",
  "pitm",
  "pixi",
  "pm2t",
  "pmax",
  "pred",
  "prol",
  "prtp",
  "pss1",
  "rap ",
  "rash",
  "recr",
  "refs",
  "res3",
  "resa",
  "resa",
  "resf",
  "resm",
  "resp",
  "ress",
  "rest",
  "resu",
  "resv",
  "rinf",
  "rloc",
  "rm2t",
  "roll",
  "rref",
  "rrgn",
  "rror",
  "rrtp",
  "rsrp",
  "rssr",
  "rtcp",
  "rtp ",
  "rtp ",
  "sabt",
  "saio",
  "saiz",
  "sap ",
  "sbas",
  "sbgp",
  "sbtt",
  "scal",
  "schi",
  "schm",
  "scif",
  "scnm",
  "scrb",
  "sdep",
  "sdp ",
  "sdtp",
  "segr",
  "seii",
  "sgpd",
  "sinf",
  "skip",
  "sm2t",
  "smhd",
  "snim",
  "snro",
  "snut",
  "spid",
  "spli",
  "spor",
  "srat",
  "srpp",
  "srtp",
  "sstl",
  "stbl",
  "stco",
  "stcp",
  "stdp",
  "ster",
  "ster",
  "sthd",
  "stmg",
  "stmi",
  "stpp",
  "strd",
  "stri",
  "strk",
  "stsa",
  "stsc",
  "stsd",
  "stsg",
  "stsh",
  "stss",
  "stsz",
  "stti",
  "stts",
  "stvi",
  "stxt",
  "stz2",
  "subp",
  "subs",
  "subs",
  "subt",
  "sulm",
  "svdr",
  "svip",
  "svpr",
  "swfr",
  "swtc",
  "swtk",
  "swto",
  "sync",
  "tbas",
  "tbas",
  "tele",
  "thmb",
  "thmb",
  "tibr",
  "tims",
  "tiri",
  "tkhd",
  "tlou",
  "tmax",
  "tmin",
  "tols",
  "totl",
  "tpay",
  "tpyl",
  "trak",
  "tran",
  "tref",
  "trep",
  "trex",
  "trgr",
  "trif",
  "trpy",
  "tsas",
  "tscl",
  "tsel",
  "tsro",
  "tssy",
  "tstb",
  "tsyn",
  "ttyp",
  "txtC",
  "tyco",
  "udes",
  "udta",
  "uriI",
  "urim",
  "url ",
  "urn ",
  "uuid",
  "vdep",
  "vipr",
  "vlab",
  "vmhd",
  "vopi",
  "vplx",
  "vref",
  "vreg",
  "vttC",
  "vvcN",
  "vvcb",
  "vwdi",
  "vwid",
  "wbbr",
  "wbbr",
  "wbbr",
  "wvtt",
  "xml ",
]
`;
